lib/promiscuous/kafka/poseidon.rb

```diff
@@ -0,0 +1,149 @@
+    zk.close
+    broker_pool.close
+
+    true
+  end
+
+  attr_accessor :connection, :connection_lock
+
+  def initialize
+    # The poseidon socket doesn't like when multiple threads access to it apparently
+    @connection_lock = Mutex.new
+  end
+
+  def new_connection
+    # TODO: client id needs to be unique across the system, use hostname+rand?
+    client_id = ['promiscuous', Promiscuous::Config.app, Poseidon::Cluster.guid].join('.')
+    connection = ::Poseidon::Producer.new(Promiscuous::Config.kafka_hosts, client_id,
+                                          :type => :sync,
+                                          :compression_codec => :none,
+                                          :metadata_refresh_interval_ms => 600_000,
+                                          :max_send_retries => 10,
+                                          :retry_backoff_ms => 100,
+                                          :required_acks => 1,
+                                          :ack_timeout_ms => 1000,
+                                          :socket_timeout_ms => Promiscuous::Config.socket_timeout)
+    connection
+  end
+
+  def connect
+    @connection = new_connection
+    # Poseidon.logger = Logger.new(STDOUT).tap { |l| l.level = 0 }
+  end
+
+  def disconnect
+    @connection_lock.synchronize do
+      return unless connected?
+      @connection.shutdown
+      @connection = nil
+    end
+  end
+
+  #TODO: broken
+  def connected?
+    !!@connection.try(:connected?)
+  end
+
+  def raw_publish(options)
+    ok = false
+    10.times do
+      ok = @connection.send_messages([Poseidon::MessageToSend.new(options[:topic], options[:payload], options[:key])])
+      break if ok
+      sleep(1)
+    end
+    raise "Unable to send messages" if !ok
+  end
+
+  def publish(options={})
+    @connection_lock.synchronize do
+      raw_publish(options)
+      options[:on_confirm].call if options[:on_confirm]
+      Promiscuous.debug "[publish] [kafka] #{options[:topic]}/#{options[:topic_key]} #{options[:payload]}"
+    end
+  rescue Exception => e
+    Promiscuous.warn("[publish] Failure publishing to kafka #{e}\n#{e.backtrace.join("\n")}")
+    e = Promiscuous::Error::Publisher.new(e, :payload => options[:payload])
+
+    if options[:async]
+      Promiscuous::Config.error_notifier.call(e)
+    else
+      raise e
+    end
+  end
+
+  module Subscriber
+    def subscribe(topic)
+      @consumer = ::Poseidon::ConsumerGroup.new(Promiscuous::Config.app,
+                                                Promiscuous::Config.kafka_hosts,
+                                                Promiscuous::Config.zookeeper_hosts,
+                                                topic, :trail => Promiscuous::Config.test_mode, :max_wait_ms => 10)
+    end
+
+    def fetch_and_process_messages(&block)
+      Promiscuous.debug "[kafka] Fetching messages topic:#{@consumer.topic} #{Thread.current}"
+      @consumer.fetch(:commit => false) do |partition, payloads|
```

Comment:
not sure what you mean, example? I thought we were supposed to line up all of the arguments with the first, which is what I had done here.